compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -319,7 +319,37 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     }
                 }
                 ItemKind::TyAlias(self_ty, _) => icx.to_ty(self_ty),
-                ItemKind::Impl(hir::Impl { self_ty, .. }) => icx.to_ty(*self_ty),
+                ItemKind::Impl(
+                    hir::Impl { self_ty, .. }
+                ) => {
+                    struct MyVisitor(bool);
+                    impl<'v> hir::intravisit::Visitor<'v> for MyVisitor {
+                        fn visit_ty(&mut self, t: &'v Ty<'v>) {
+                            if matches!(
+                                &t.kind,
+                                TyKind::Path(hir::QPath::Resolved(
+                                    _,
+                                    Path {
+                                        res: hir::def::Res::SelfTyAlias { .. },
+                                        ..
+                                    },
+                                ))
+                            ) {
+                                self.0 = true;
+                                return;
+                            }
+                            hir::intravisit::walk_ty(self, t);
+                        }
+                    }
+
+                    let mut my_visitor = MyVisitor(false);
+                    my_visitor.visit_ty(self_ty);
+
+                    match my_visitor.0 {
+                        true => { tcx.sess.emit_err(crate::errors::SelfInImplSelf { span: self_ty.span}); tcx.ty_error() },
```

Comment:
It'd be nice if we got a span back from the `Visitor` to point at the *exact* place where the `Self` is used, instead of pointing at the whole `Ty` place, because it can be things like `dyn ToNbt<Self>`, or much longer and it might be confusing if we point at the whole thing in those cases.

Reply:
You can be fancy and see if the two are different and have a secondary span pointing at the whole type, but that's not entirely necessary.